File: boj_10866.cpp
#include <deque>
#include <iostream>
#include <string>
using namespace std;

int main() {
  int n = 0;
  cin >> n;
  cin.ignore();

  deque<int> dq;

  for (int i = 0; i < n; ++i) {
    string order;
    getline(cin, order);

    if (order == "front") {
      if (dq.empty()) {
        cout << -1 << "\n";
      } else {
        cout << dq.front() << "\n";
      }
    }

    else if (order == "back") {
      if (dq.empty()) {
        cout << -1 << "\n";
      } else {
        cout << dq.back() << "\n";
      }
    } else if (order[1] ==
               'u') {  // check abt order == push
                       // order[1] is char type data, so only can compare using
                       // ' ' if you use " " for compare order[1], there will be
                       // warnings with red line
      string backup = order;
      if (order[5] == 'f') {                // check push_'f'ront
        int x = stoi(backup.erase(0, 10));  // stoi mean "string to integer"
        dq.push_front(x);
      } else {  // push_back
        int x = stoi(backup.erase(0, 9));
        dq.push_back(x);
      }
    }

    else if (order[1] == 'o') {  // check abt order == pop
      if (order[4] == 'f') {     // check pop_'f'ront
        if (dq.empty()) {
          cout << -1 << "\n";
        } else {
          cout << dq.front() << "\n";
          dq.pop_front();
        }
      } else {  // check pop_back
        if (dq.empty()) {
          cout << -1 << "\n";
        } else {
          cout << dq.back() << "\n";
          dq.pop_back();
        }
      }
    } else if (order == "size") {
      cout << dq.size() << "\n";
    } else if (order == "empty") {
      if (dq.empty()) {
        cout << 1 << "\n";
      } else {
        cout << 0 << "\n";
      }
    }
  }
}

File: boj_10866.py
import sys
from collections import deque


class bojDeQue:
    def __init__(self):
        self.deque = deque()
        self.len = 0

    def push_front(self, x):
        self.deque.appendleft(x)
        self.len += 1

    def push_back(self, x):
        self.deque.append(x)
        self.len += 1

    def pop_front(self):
        if self.len == 0:
            return -1
        else:
            self.len -= 1
            return self.deque.popleft()

    def pop_back(self):
        if self.len == 0:
            return -1
        else:
            self.len -= 1
            return self.deque.pop()

    def size(self):
        return self.len

    def empty(self):
        if self.len == 0:
            return 1
        else:
            return 0

    def front(self):
        if self.len == 0:
            return -1
        else:
            return self.deque[0]

    def back(self):
        if self.len == 0:
            return -1
        else:
            return self.deque[-1]


n = int(input())
deQ = bojDeQue()

while n > 0:
    n -= 1
    order = list(sys.stdin.readline().split())
    if order[0] == "push_front":
        deQ.push_front(order[1])
    elif order[0] == "push_back":
        deQ.push_back(order[1])
    elif order[0] == "pop_front":
        print(deQ.pop_front())
    elif order[0] == "pop_back":
        print(deQ.pop_back())
    elif order[0] == "size":
        print(deQ.size())
    elif order[0] == "empty":
        print(deQ.empty())
    elif order[0] == "front":
        print(deQ.front())
    elif order[0] == "back":
        print(deQ.back())
